onnx/checker.cc

```diff
@@ -226,6 +228,160 @@ void check_tensor(const TensorProto& tensor, const CheckerContext& ctx) {
+void check_sparse_tensor_indices_2(
+    const TensorProto& indices,
+    const SparseTensorProto& sparse_tensor_proto,
+    size_t nnz) {
+  int dense_rank = sparse_tensor_proto.dims_size();
+  if (static_cast<size_t>(indices.dims(0)) != nnz)
+    fail_check(
+        "Sparse tensor indices (",
+        indices.name(),
+        ") first dimension size does not equal NNZ.");
+  if (indices.dims(1) != dense_rank)
+    fail_check(
+        "Sparse tensor indices (",
+        indices.name(),
+        ") second dimension size does not match rank of tensor.");
+
+  // Check if indices appear in ascending order, and if they have valid
+  // values.
+  const std::vector<int64_t> index_data = ParseData<int64_t>(&indices);
+
+  int64_t prev_index = -1;
+  for (size_t i = 0; i < nnz; ++i) {
+    int64_t curr_index = 0; // linearized index of i-th value
+    for (int j = 0; j < dense_rank; ++j) {
+      auto index_ij = index_data[i * dense_rank + j];
+      if ((index_ij < 0) || (index_ij >= sparse_tensor_proto.dims(j)))
+        fail_check(
+            "Sparse tensor (",
+            indices.name(),
+            ") index value at position [",
+            i,
+            ",",
+            j,
+            "] out of range.");
+      curr_index = curr_index * sparse_tensor_proto.dims(j) + index_ij;
+    }
+    if (curr_index <= prev_index) {
+      fail_check(
+          "Sparse tensor (",
+          indices.name(),
+          ") index value at position [",
+          i,
+          "] not in lexicographic sorted order.");
+    }
+    prev_index = curr_index;
+  }
+}
+
+void check_sparse_tensor(
+    const SparseTensorProto& sparse_tensor_proto,
+    const CheckerContext& ctx) {
+  enforce_has_field(sparse_tensor_proto, values);
+
+  const TensorProto& values = sparse_tensor_proto.values();
+  check_tensor(values, ctx);
+
+  // values must be a tensor of shape [NNZ]
+  // Currently we restrict the value associated with a particular index-tuple
+  // to be a single value. In the future, if there is a requirement,
+  // we may extend this to permit the value to be a "sub-tensor", in which
+  // case values will have dimension > 1.
+  if (values.dims_size() != 1)
+    fail_check("Sparse tensor values (", values.name(), ") must have rank 1.");
+  size_t nnz = static_cast<size_t>(values.dims(0));
+
+  int dense_rank = sparse_tensor_proto.dims_size();
+  if (dense_rank == 0) {
+    // TODO: Should we add a name field for a sparse-tensor-proto?
+    // Currently, values has a name, but message may be a bit confusing.
+    fail_check(
+        "Sparse tensor (", values.name(), ") must have a dense-rank > 0");
+  }
+  for (int i = 0; i < dense_rank; ++i) {
+    if (sparse_tensor_proto.dims(i) <= 0)
+      fail_check(
+          "Sparse tensor (", values.name(), ") dimensions are not positive.");
+  }
+
+  if (sparse_tensor_proto.has_indices()) {
+    // We keep the representation simple for now.
+    // indices must be a tensor of shape [NNZ, dense_rank].
```

Comment:
`Must`? I saw two formats below. One uses linearized indexing and the other is the same as what described here.
```